core/editor/src/tool/tools/select.rs

```diff
@@ -162,6 +168,36 @@ impl Fsm for SelectToolFsmState {
 
 					Ready
 				}
+				(_, AlignLeft) => {
+					align_selected(document, responses, AlignDimension::X, AlignAggregate::Min);
+
+					self
+				}
+				(_, AlignHorizontalCenter) => {
+					align_selected(document, responses, AlignDimension::X, AlignAggregate::Average);
+
+					self
+				}
+				(_, AlignRight) => {
+					align_selected(document, responses, AlignDimension::X, AlignAggregate::Max);
+
+					self
+				}
+				(_, AlignTop) => {
+					align_selected(document, responses, AlignDimension::Y, AlignAggregate::Min);
+
+					self
+				}
+				(_, AlignVerticalCenter) => {
+					align_selected(document, responses, AlignDimension::Y, AlignAggregate::Average);
+
+					self
+				}
+				(_, AlignBottom) => {
+					align_selected(document, responses, AlignDimension::Y, AlignAggregate::Max);
+
+					self
+				}
```

Comment:
I'm surprised we would need to add those button actions to the select tool's FSM. Why doesn't the button just send a message directly to the document? I'd be concerned that this approach of involving the tool's FSM could cause additional complication if it isn't necessary. But I am also not the most familiar with this part of the code, so if you think it's necessary, I'd be delighted to learn why to help me understand this code better.

Reply:
This is the approach that @TrueDoctor and I converged on. A big advantage of it is that button message handling can depend on state (in a semantically clear way), which we think may be a desirable use case. Similarly, button messages could easily act on state, e.g. aborting a drag action if a certain action is done (maybe by a keyboard shortcut). It also keeps all of a tool's message handling in one place, and keeps every message within the FSM metaphor (button presses generally just being self-transitions).

I did substantially cut down on the code by making this a tuple variant, though.